coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+    return false;
+  }
+}
+```
+> **Note:** In addition to the naming convention this price doesn't check if the price number is > 0 to verify that they aren't adding funcs into the wallet buy "paying" for negative values.
+
+The above example illustrates in a way that the function name is no longer discriptive of what the code does. It also starts to blend into a section covered later [Single Responsibility Pattern](#single-responsibility-pattern), but first let's cover the naming convention. A way to make the function more discriptive would be:
+```typescript
+...
+payFromWalletOrAttemptToLendMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now the function name mirrors to an extend the behavior of the function, but in a way the function is still a bit wrong. We return a boolean to see if the purchase was accepted, so a better name might have been.
+```typescript
+...
+successfullyPaidFromWalletOrLentMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now in addition to describing what happens inside the function it also matches with the return type of the function.
+
+# Arrays
+
+### [Array Instantiation](#array-instantiation)
+For array instantiation you should use the literal syntax instead of `new` constructor.
+```typescript
+// const items: number[] = new Array(); Don't do this.
+const items: number[] = [];
+```
+
+### [Array Operations](#array-operations)
+For operations on arrays use push etc instead of direct assignment.
+```typescript
+const items: number[] = [];
+// items[items.length] = 14; No, don't do this.
+items.push(14);
+```
+
+# Functions
+
+### [Expression Vs Declaration](#expression-vs-declaration)
+When defining a function we prefer function declaration over function expression, and when dealing with assigned functions we prefer using using an anonymous [arrow function](#arrow-functions) expression.
+```typescript
+function helloWorld(): void {} // Yes
+const helloWorld = function(): Void => {}; // No
+const helloWorld = (): void => {}; // No
+```
+> **Note:** Performance based on numbers from: https://jsperf.com/function-vs-arrow-function/14.
+
+> **Warning:** Declaring a function in a non-function block (if, while etc) has had a history of having different behavior across browsers. It is recommended to assign the function to a variable instead.
+
+### [Don't use Arguments](#dont-use-arguments)
+To clarify, your functions can have arguments, but you should never name a parameter `arguments` as this will take precedence over the similarly named object on the function scope, and you don't want to do that.
+
+### [Default Parameters](#default-parameters)
+Instead of having mutating function arguments, consider using default parameters.
+```typescript
+// Worst implementation
+function handleThings(opts): void {
+  // No! We shouldn't mutate function arguments.
+  // Double bad: if opts is falsy it'll be set to an object which may
+  // be what you want but it can introduce subtle bugs.
+  opts = opts || {};
```

Comment:
This is a pattern we use quite extensively... What do you propose we do with the usage we have? I think it's a quite nice pattern to use with options-objects, but I agree that falsy values is a bit risky. Maybe we should be explicit about matching undefined?

Reply:
Well, as proposed further down is to rather add the default to the function variable, and yes, that means if we agree on this we should change our old code to match the guideline.